cpp/cmake_modules/ThirdpartyToolchain.cmake

```diff
@@ -3865,6 +3902,212 @@ if(ARROW_ORC)
+      )
+    endif()
+    add_library(opentelemetry-cpp::${_OPENTELEMETRY_LIB} STATIC IMPORTED)
+    set_target_properties(opentelemetry-cpp::${_OPENTELEMETRY_LIB}
+                          PROPERTIES IMPORTED_LOCATION ${_OPENTELEMETRY_STATIC_LIBRARY})
+    list(APPEND OPENTELEMETRY_BUILD_BYPRODUCTS ${_OPENTELEMETRY_STATIC_LIBRARY})
+    list(APPEND OPENTELEMETRY_LIBRARIES opentelemetry-cpp::${_OPENTELEMETRY_LIB})
+  endforeach()
+
+  set(OPENTELEMETRY_CMAKE_ARGS
+      ${EP_COMMON_TOOLCHAIN}
+      "-DCMAKE_INSTALL_PREFIX=${OPENTELEMETRY_PREFIX}"
+      "-DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}"
+      -DCMAKE_INSTALL_LIBDIR=lib
+      "-DCMAKE_CXX_FLAGS=${EP_CXX_FLAGS}"
+      -DBUILD_TESTING=OFF
+      -DWITH_EXAMPLES=OFF)
+
+  set(OPENTELEMETRY_PREFIX_PATH_LIST)
+  # Don't specify the DEPENDS unless we actually have dependencies, else
+  # Ninja/other build systems may consider this target to always be dirty
+  set(_OPENTELEMETRY_DEPENDENCIES)
+  add_custom_target(opentelemetry_dependencies)
+
+  if(ARROW_WITH_OPENTELEMETRY)
+    set(_OPENTELEMETRY_DEPENDENCIES "opentelemetry_dependencies")
+    list(APPEND ARROW_BUNDLED_STATIC_LIBS ${OPENTELEMETRY_LIBRARIES})
+    list(APPEND OPENTELEMETRY_PREFIX_PATH_LIST ${NLOHMANN_JSON_PREFIX})
+    set(OPENTELEMETRY_CMAKE_ARGS ${OPENTELEMETRY_CMAKE_ARGS} -DWITH_OTLP=ON
+                                 -DWITH_OTLP_HTTP=ON -DWITH_OTLP_GRPC=OFF)
+
+    # OpenTelemetry with OTLP enabled requires Protobuf definitions from a
+    # submodule. This submodule path is hardcoded into their CMake definitions,
+    # and submodules are not included in their releases. Add a custom build step
+    # to download and extract the Protobufs.
+
+    # Adding such a step is rather complicated, so instead: create a separate
+    # ExternalProject that just fetches the Protobufs, then add a custom step
+    # to the main build to copy the Protobufs.
+    externalproject_add(opentelemetry_proto_ep
+                        ${EP_LOG_OPTIONS}
+                        URL_HASH "SHA256=${ARROW_OPENTELEMETRY_PROTO_BUILD_SHA256_CHECKSUM}"
+                        URL ${OPENTELEMETRY_PROTO_SOURCE_URL}
+                        BUILD_COMMAND ""
+                        CONFIGURE_COMMAND ""
+                        INSTALL_COMMAND ""
+                        EXCLUDE_FROM_ALL OFF)
+
+    add_dependencies(opentelemetry_dependencies nlohmann_json_ep opentelemetry_proto_ep)
+    if(gRPC_SOURCE STREQUAL "BUNDLED")
+      # TODO: opentelemetry-cpp::proto doesn't declare a dependency on gRPC, so
+      # even if we provide the location of gRPC, it'll fail to compile.
+      message(FATAL_ERROR "ARROW_WITH_OPENTELEMETRY cannot be configured with gRPC_SOURCE=BUNDLED. "
+                          "See https://github.com/open-telemetry/opentelemetry-cpp/issues/1045"
+      )
+    endif()
+  else()
+    set(OPENTELEMETRY_CMAKE_ARGS ${OPENTELEMETRY_CMAKE_ARGS} -DWITH_API_ONLY=ON)
```

Comment:
Could you use `list(APPEND ...)`?

Reply:
Sorry, I didn't fully change OpenTelemetry to be a completely optional component. I've reorganized things such that there's only a singular ARROW_WITH_OPENTELEMETRY check. As a consequence this branch is no longer needed in the first place.